shogun-lib/src/main/java/de/terrestris/shogun/lib/service/GroupService.java

```diff
@@ -0,0 +1,119 @@
+
+@Service
+public class GroupService extends BaseService<GroupRepository, Group> {
+
+    @Autowired
+    KeycloakUtil keycloakUtil;
+
+    @Autowired
+    UserRepository userRepository;
+
+    @PostFilter("hasRole('ROLE_ADMIN') or hasPermission(filterObject, 'READ')")
+    @Transactional(readOnly = true)
+    @Override
+    public List<Group> findAll() {
+        List<Group> groups = (List<Group>) repository.findAll();
+
+        for (Group group : groups) {
+            this.setTransientKeycloakRepresentations(group);
+        }
+
+        return groups;
+    }
+
+    @PostFilter("hasRole('ROLE_ADMIN') or hasPermission(filterObject, 'READ')")
+    @Transactional(readOnly = true)
+    @Override
+    public List<Group> findAllBy(Specification specification) {
+        List<Group> groups = (List<Group>) repository.findAll(specification);
+
+        for (Group group : groups) {
+            this.setTransientKeycloakRepresentations(group);
+        }
+
+        return groups;
+    }
+
+    @PostAuthorize("hasRole('ROLE_ADMIN') or hasPermission(returnObject.orElse(null), 'READ')")
+    @Transactional(readOnly = true)
+    @Override
+    public Optional<Group> findOne(Long id) {
+        Optional<Group> group = repository.findById(id);
+
+        if (group.isPresent()) {
+            group = Optional.of(this.setTransientKeycloakRepresentations(group.get()));
+        }
+
+        return group;
+    }
+
+    @PostFilter("hasRole('ROLE_ADMIN') or hasPermission(filterObject, 'READ')")
+    @Transactional(readOnly = true)
+    public List<Group> findByUser(User user) {
+        List<Group> groups = new ArrayList<>();
+
+        List<GroupRepresentation> keycloakGroups = keycloakUtil.getUserGroups(user);
+
+        for (GroupRepresentation keycloakGroup : keycloakGroups) {
+            Optional<Group> group = repository.findByKeycloakId(keycloakGroup.getId());
+            if (group.isPresent()) {
+                group.get().setKeycloakRepresentation(keycloakGroup);
+                groups.add(group.get());
+            }
+        }
+
+        return groups;
+    }
+
+    public GroupRepresentation findByKeycloakId(String keycloakId) {
+        GroupResource groupResource = keycloakUtil.getGroupResource(keycloakId);
+
+        return groupResource.toRepresentation();
+    }
+
+    public List<User> getGroupMembers(String id) {
+        GroupResource groupResource = keycloakUtil.getGroupResource(id);
+        List<UserRepresentation> groupMembers = groupResource.members();
+
+        ArrayList<User> users = new ArrayList<>();
+        for (UserRepresentation groupMember : groupMembers) {
+            Optional<User> user = userRepository.findByKeycloakId(groupMember.getId());
+            if (user.isPresent()) {
+                user.get().setKeycloakRepresentation(groupMember);
+                users.add(user.get());
+            }
+        }
+
+        return users;
+    }
+
+    private Group setTransientKeycloakRepresentations(Group group) {
```

Comment:
Is this some kind of semi-automated update process? I'm not sure if i like that this is called on every service method. Is this really necessary?

Reply:
This only sets the transient field (the group details saved in Keycloak) in the group instance.